File: Middle-of-LinkedList/terceira.cpp
#include <iostream>
#include <math.h>
#include <stdio.h>
class ListNode {
public:
    int val;
    ListNode* next;
    ListNode() {
        this->val = 0;
        this->next = nullptr;
    }
    ListNode(int val) {
        this->val = val;
        this->next = nullptr;
    }
    ListNode(int val, ListNode* next) {
        this->next = next;
        this->val = val;
    }
};

void printListNode(ListNode* node) {
    while (node != nullptr) {
        std::cout << node->val << " ";
        node = node->next;
    }
    std::cout << "\n";
}

// Naive
ListNode* middleNodeNaive(ListNode* head) {
    ListNode* pointer = head;
    int size = 0;
    while (pointer != nullptr && pointer->next != nullptr) {
        pointer = pointer->next;
        size++;
    }
    pointer = head;
    for (int i = 0; i < ceil((double)size / 2); i++)
        pointer = pointer->next;

    return pointer;
}

// Optimal
ListNode* middleNode(ListNode* head) {
    ListNode* slowPointer = head;
    ListNode* fastPointer = head;
    while(fastPointer != nullptr && fastPointer->next != nullptr){
        fastPointer = fastPointer->next->next;
        slowPointer = slowPointer->next;
    }
    return slowPointer;
}

int main() {
    printListNode(middleNode(
        new ListNode(1, new ListNode(2, new ListNode(3)))));

    return 0;
}

File: Middle-of-LinkedList/terceira.py
from math import ceil

from sympy import im

class ListNode:
    val:int
    next_:'ListNode'
    def __repr__(self) -> str:
        cur = self
        result = ""
        while cur != None:
            result += f"{cur.val} "
            cur = cur.next_
        return result

    def __init__(self, val:int = 0, next: 'ListNode' = None) -> None:
        self.val = val
        self.next_ = next

def middleLinkedListNaive(head:'ListNode') -> 'ListNode':
    size = 0
    pointer = head
    while pointer != None:
        pointer = pointer.next_
        size+= 1
    middle = head
    for i in range(ceil(size//2)):
        middle = middle.next_
    return middle

def middleLinkedListFast(head:'ListNode') -> 'ListNode':
    slowPointer = head
    fastPointer = head
    while fastPointer != None and fastPointer.next_ != None:
        slowPointer = slowPointer.next_
        fastPointer = fastPointer.next_.next_
    return slowPointer

impar = ListNode(1, ListNode(2, ListNode(3)))
par   = ListNode(1, ListNode(2, ListNode(3, ListNode())))
print(middleLinkedListFast(impar))
print(middleLinkedListFast(par))